src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt

```diff
@@ -29,20 +30,45 @@ import io.grpc.StatusException
  */
 class CosmasGoogleCloudService(private val bucketName: String,
                                private val storage: Storage = StorageOptions.getDefaultInstance().service) : CosmasGrpc.CosmasImplBase() {
+    private val fileBuffer = ConcurrentHashMap<String, ConcurrentHashMap<String, ByteString>>().withDefault { ConcurrentHashMap() }
 
     override fun createVersion(request: CosmasProto.CreateVersionRequest,
                                responseObserver: StreamObserver<CosmasProto.CreateVersionResponse>) {
         println("Get request for create new version of file # ${request.fileId}")
+        val project = this.fileBuffer.getValue(request.projectId)
+        project[request.fileId] = request.file
+        this.fileBuffer[request.projectId] = project
+
+        val response = CosmasProto.CreateVersionResponse
+                .newBuilder()
+                .build()
+        responseObserver.onNext(response)
+        responseObserver.onCompleted()
+    }
+
+    override fun commitVersion(request: CosmasProto.CommitVersionRequest, responseObserver: StreamObserver<CosmasProto.CommitVersionResponse>) {
+        println("Get request for commit last version of files in project # ${request.projectId}")
+        val project = this.fileBuffer[request.projectId]
+
+        if (project == null) {
+            val status = Status.INVALID_ARGUMENT.withDescription(
+                    "There is no project in buffer with project id ${request.projectId}")
+            println(status.description)
+            responseObserver.onError(StatusException(status))
+            return
+        }
         try {
-            val blob = this.storage.create(
-                    BlobInfo.newBuilder(this.bucketName, request.fileId).build(),
-                    request.file.toByteArray())
-            println("Generation of created file: ${blob.generation}")
+            project.forEach { (fileId, file) ->
```

Comment:
This is the case when ConcurrentHashMap is useful. Regular HashMap throws ConcurrentModificationException if map is modified concurrently with iteration. It can be modified if someone calls createVersion concurrently with commitVersion, that's why I asked to do something here. But it doesn't mean that you don't need other kinds of synchronization.